### PR TITLE
tweak(sync): NASS-1730: Update sync behaviour based on is_sensitive column

### PR DESCRIPTION
**Pull request by tcodling**
### Changes

In the epic we added a new `is_sensitive` column to facilities to support some new sync behaviour. This pr implements that behaviour. 

Included is the following:
- Update of `buildEncounterLinkedLookup` filter to add a `facilityId` to the lookup record if its from a `is_sensitive` facility
  - This should mean that these records only sync down to that facility and no others
- Swap out all models for encounter linked data to ensure they all use this sync filter
- Move any custom arguments/overrides to `buildEncounterLinkedLookup`
- Very large test suite checking that each data type does not sync away from a sensitive facility

### Deploys

- [x] **Deploy to Tamanu Internal** 

### Tests

- [ ] **Run E2E Tests** 

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

### CODE REVIEW

**Comment by tcodling** [comment]
bugbot run

**Comment by tcodling** [comment]
bugbot run

File: packages/central-server/__tests__/sync/CentralSyncManager.syncLookup.test.js [hunk 1]
```diff
@@ -712,17 +712,11 @@ describe('Sync Lookup data', () => {
...
-      const expectedFacility = ['appointments', 'appointment_schedules'].includes(model.tableName)
-        ? facility.id
-        : null;
-
       expect(syncLookupRecord.dataValues).toEqual(
         expect.objectContaining({
           recordId: expect.anything(),
           recordType: model.tableName,
           patientId: expect.anything(),
-          facilityId: expectedFacility,
```
Comment (tcodling) [inline packages/central-server/__tests__/sync/CentralSyncManager.syncLookup.test.js:725]: Now facilityId is optional for all encounter linked models so this check is no longer valid

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js [hunk 2]
```diff
@@ -2838,4 +2838,923 @@ describe('CentralSyncManager', () => {
       });
     });
   });
+
+  describe('facility sensitivity sync filtering', () => {
```
Comment (tcodling) [inline packages/central-server/__tests__/sync/CentralSyncManager.test.js:2842]: I worry a little about the performance of this describe block. It adds about a minute or more to the test suite on my local due to the amount of syncing/updating lookup table it does. 

I did look into trying to combine the test setup/syncing but i found negligible increases in performance and it became much harder to read. Am open to ideas here otherwise thinking I will create a card to work on improving performance if I have time.

File: packages/database/src/sync/buildEncounterLinkedLookupFilter.ts [hunk 3]
```diff
@@ -1,12 +1,55 @@
...
+export function addSensitiveFacilityIdIfApplicable() {
+  return `
+    CASE
+      WHEN facilities.is_sensitive = TRUE THEN facilities.id
+      ELSE NULL
+    END
+  `;
+}
+
+export function buildEncounterLinkedLookupFilter(model: typeof Model, options?: Options) {
```
Comment (tcodling) [inline packages/database/src/sync/buildEncounterLinkedLookupFilter.ts:32]: Super open to suggestions on this one. I essentially had to move all the weird custom logic, arguments from the individual sync filters into here so its a bit messier than I would like ideally.

File: packages/database/src/sync/buildEncounterLinkedLookupFilter.ts [hunk 4]
```diff
@@ -1,12 +1,55 @@
...
-    select: buildSyncLookupSelect(model, {
-      patientId: 'encounters.patient_id',
-    }),
-    joins: buildEncounterLinkedSyncFilterJoins([model.tableName, 'encounters']),
-  };
+export type JoinConfig = {
+  tableName: string;
+  columnName: string;
+  joinType?: 'LEFT' | 'INNER';
+};
```
Comment (tcodling) [inline packages/database/src/sync/buildEncounterLinkedLookupFilter.ts:10]: Follow sequilze syntax if possible (use type if possible)

**Comment by tcodling** [comment]
bugbot run

**Comment by tcodling** [comment]
Determinism failing but I think this should resolve once merged to the epic. Then I will start a proper dev test 🙏

File: packages/database/src/sync/buildEncounterLinkedLookupFilter.ts [hunk 5]
```diff
@@ -1,12 +1,55 @@
...
+  extraSelects?: Record<string, string>; // extra selects to attach to the lookup select
+  extraJoins?: JoinConfig[]; // extra joins needed to traverse between this model and the encounters table
+};
+
+/**
+ * Helper function to determine if a facility_id should be populated in sync lookup
+ * Only populates facility_id when the encounter is from a sensitive facility
+ * This ensures sensitive encounters are only synced to their originating facility
+ */
+export function addSensitiveFacilityIdIfApplicable() {
```
Comment (edmofro) [inline packages/database/src/sync/buildEncounterLinkedLookupFilter.ts:18]: Could be a const rather than a function - all it does is return the same text every time. Funnily enough I think that will make the usage site less readable, but it is the normal thing to do.

File: packages/database/src/sync/buildEncounterLinkedLookupFilter.ts [hunk 6]
```diff
@@ -1,12 +1,55 @@
...
+import {
+  buildEncounterLinkedSyncFilterJoins,
+  type JoinConfig,
+} from './buildEncounterLinkedSyncFilter';
 import { buildSyncLookupSelect } from './buildSyncLookupSelect';
 import type { Model } from '../models/Model';
 
+export type Options = {
+  extraSelects?: Record<string, string>; // extra selects to attach to the lookup select
+  extraJoins?: JoinConfig[]; // extra joins needed to traverse between this model and the encounters table
```
Comment (edmofro) [inline packages/database/src/sync/buildEncounterLinkedLookupFilter.ts:10]:
```suggestion
  joinsToEncounters?: JoinConfig[]; // extra joins needed to traverse between this model and the encounters table
```
Matching naming in params below

Comment (edmofro) [reply packages/database/src/sync/buildEncounterLinkedLookupFilter.ts:10]: Oh yo is this used??

File: packages/database/src/sync/buildEncounterLinkedSyncFilter.ts [hunk 7]
```diff
@@ -1,18 +1,36 @@
...
-    `,
-    )
+    .map((table, i) => {
+      const currentTable = isString(tablesToTraverse[i])
+        ? tablesToTraverse[i]
+        : tablesToTraverse[i]?.model.tableName;
+
+      const joinTable = isString(table) ? table : table.model.tableName;
+      const joinColumn = isString(table) ? `${Utils.singularize(table)}_id` : table.joinColumn;
+      const joinType = isObject(table) && table.required ? 'INNER' : 'LEFT';
```
Comment (edmofro) [inline packages/database/src/sync/buildEncounterLinkedSyncFilter.ts:23]: Is `required` the same key sequelize uses?

File: packages/database/src/sync/buildEncounterLinkedSyncFilter.ts [hunk 8]
```diff
@@ -1,18 +1,36 @@
...
+      return `
+        ${joinType} JOIN ${joinTable} ON ${currentTable}.${joinColumn} = ${joinTable}.id
+      `;
+    })
     .join('\n');
 }
 
 export function buildEncounterLinkedSyncFilter(
-  tablesToTraverse: string[], // e.g. [ 'survey_response_answers', 'survey_responses', 'encounters'] to traverse up from survey_response_answers
+  tablesToTraverse: JoinConfig[], // e.g. [ 'survey_response_answers', 'survey_responses', 'encounters'] to traverse up from survey_response_answers
```
Comment (edmofro) [inline packages/database/src/sync/buildEncounterLinkedSyncFilter.ts:33]: Could add an object to the e.g. in the comment

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js [hunk 9]
```diff
@@ -2838,4 +2838,965 @@ describe('CentralSyncManager', () => {
...
+        maxRecordsPerSnapshotChunk: DEFAULT_MAX_RECORDS_PER_SNAPSHOT_CHUNKS,
+      },
+    };
+
+    beforeEach(async () => {
+      patient = await models.Patient.create(fake(models.Patient));
+      practitioner = await models.User.create(fake(models.User));
+      // Reset modules to ensure fresh imports
+      jest.resetModules();
+      // The config override actually doesnt apply to snapshotOutgoingChanges which uses the default.json
```
Comment (edmofro) [inline packages/central-server/__tests__/sync/CentralSyncManager.test.js:2864]:
```suggestion
      // The config override actually doesn't apply to snapshotOutgoingChanges which uses the default.json, so we need to mock it directly here
```

Comment (edmofro) [reply packages/central-server/__tests__/sync/CentralSyncManager.test.js:2864]: What do you mean it uses `default.json` though?

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js [hunk 10]
```diff
@@ -2838,4 +2838,965 @@ describe('CentralSyncManager', () => {
...
+            `No records found for record type ${model.tableName} in lookup table, Check the test setup!`,
+          );
+        }
+
+        expect(recordIds).not.toContain(sensitiveId);
+        expect(recordIds).toContain(nonSensitiveId);
+      };
+
+      // Basic encounter
+      it('wont sync sensitive encounters', async () => {
```
Comment (edmofro) [inline packages/central-server/__tests__/sync/CentralSyncManager.test.js:2971]:
```suggestion
      it("won't sync sensitive encounters", async () => {
```
etc. below

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js [hunk 11]
```diff
@@ -2838,4 +2838,965 @@ describe('CentralSyncManager', () => {
...
+    const lookupEnabledConfig = {
+      sync: {
+        lookupTable: {
+          enabled: true,
+        },
+        maxRecordsPerSnapshotChunk: DEFAULT_MAX_RECORDS_PER_SNAPSHOT_CHUNKS,
+      },
+    };
+
+    beforeEach(async () => {
```
Comment (edmofro) [inline packages/central-server/__tests__/sync/CentralSyncManager.test.js:2859]: I think this could be a `beforeAll`, and save a lot of time?

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js [hunk 12]
```diff
@@ -2838,4 +2838,965 @@ describe('CentralSyncManager', () => {
...
+        expect(encounterIds).not.toContain(sensitiveEncounterB.id);
+      });
+
+      it.todo(
+        'will leave historical data unsynced when the facility changes from sensitive to non-sensitive until it is edited',
+      );
+
+      it.todo(
+        'will leave historical data unsynced when the facility changes from non sensitive to sensitive',
+      );
```
Comment (edmofro) [inline packages/central-server/__tests__/sync/CentralSyncManager.test.js:3799]: 👀

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js [hunk 13]
```diff
@@ -2838,4 +2838,965 @@ describe('CentralSyncManager', () => {
...
+        );
+
+        const outgoingChanges = await centralSyncManager.getOutgoingChanges(sessionId, {});
+        const encounterIds = outgoingChanges
+          .filter(c => c.recordType === 'encounters')
+          .map(c => c.recordId);
+
+        expect(encounterIds).toContain(sensitiveEncounterA.id);
+        expect(encounterIds).not.toContain(sensitiveEncounterB.id);
+      });
```
Comment (edmofro) [inline packages/central-server/__tests__/sync/CentralSyncManager.test.js:3791]: Very quick skim through these tests, didn't really read them after the first few as they were all so similar

**Review by edmofro — CHANGES_REQUESTED** [review]
Mostly looking great! A few suggestions

File: packages/database/src/sync/buildEncounterLinkedLookupFilter.ts (hunk 6 above)
Comment (tcodling) [reply packages/database/src/sync/buildEncounterLinkedLookupFilter.ts:10]: oh it isnt!

**Comment by tcodling** [comment]
@edmofro Ready for review. If possible I was thinking of working on those edge case tests in a seperate PR so I can get Klaus started testing asap? Or do you think its a bit risky in terms of potentially having to double back on testing?

File: packages/database/src/sync/buildEncounterLinkedSyncFilter.ts (hunk 8 above)
Comment (edmofro) [reply packages/database/src/sync/buildEncounterLinkedSyncFilter.ts:33]: Only thing is that the comment makes it more obvious that `encounters` needs to be included, though that could be done through a differently structured comment

Comment (edmofro) [reply packages/database/src/sync/buildEncounterLinkedSyncFilter.ts:33]: I am totally on board with the idea of letting the type do the talking in general though!

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js [hunk 14]
```diff
@@ -2856,12 +2856,12 @@ describe('CentralSyncManager', () => {
       },
     };
 
-    beforeEach(async () => {
+    beforeAll(async () => {
```
Comment (edmofro) [inline packages/central-server/__tests__/sync/CentralSyncManager.test.js:2859]: Did this speed things up?

File: packages/database/src/sync/buildEncounterLinkedLookupFilter.ts [hunk 15]
```diff
@@ -5,32 +5,25 @@ import {
...
-
 /**
- * Helper function to determine if a facility_id should be populated in sync lookup
+ * Helper to determine if a facility_id should be populated in sync lookup
  * Only populates facility_id when the encounter is from a sensitive facility
  * This ensures sensitive encounters are only synced to their originating facility
  */
-export function addSensitiveFacilityIdIfApplicable() {
-  return `
+const addSensitiveFacilityIdIfApplicable = `
```
Comment (edmofro) [inline packages/database/src/sync/buildEncounterLinkedLookupFilter.ts:13]: Caps snake case is our usual global const convention. That's why I was meaning it would be less readable, it's definitely not as pleasant to look at, but we'd better stick with the pattern!

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js (hunk 14 above)
Comment (tcodling) [reply packages/central-server/__tests__/sync/CentralSyncManager.test.js:2859]: That actually broke everything! I have reverted for now but I will address in my cleanup card maybe?

Comment (edmofro) [reply packages/central-server/__tests__/sync/CentralSyncManager.test.js]: Sweet

**Review by edmofro — APPROVED** [review]
lgtm!
